### DISCUSSION TITLE
How does one use a custom sampler?

### DISCUSSION THREAD

**Discussion by davidji99** [open-telemetry/opentelemetry-ruby#1407] [Q&A] [answered]
Hi,

I am aware of the built-in samplers:

```
always_on - Sampler that always samples spans, regardless of the parent span’s sampling decision.

always_off - Sampler that never samples spans, regardless of the parent span’s sampling decision.

traceidratio - Sampler that samples probabalistically based on rate.

parentbased_always_on - (default) Sampler that respects its parent span’s sampling decision, but otherwise always samples.

parentbased_always_off - Sampler that respects its parent span’s sampling decision, but otherwise never samples.

parentbased_traceidratio - Sampler that respects its parent span’s sampling decision, but otherwise samples probabalistically based on rate.
```

But I'd like to understand how to use a custom sampler?

I currently have the following:

```
sampler = CustomSampler.new(sampling_probability)
trace_provider = OpenTelemetry::SDK::Trace::TracerProvider.new(sampler: sampler)
OpenTelemetry.tracer_provider = trace_provider
```

But these lines of code don't seem to be taking effect. Any guidance?

**Comment by ahayworth** [comment]
Hi @davidji99 - what you posted looks about correct, so I'm not immediately sure what the issue is. Could you give me a more substantial reproduction case, so I can dig into it a little more? Specifically I think having a snippet of code that shows the full SDK initialization you're doing along with how you're generating a trace that doesn't seem to respect your sampler - that would help a lot.

**Comment by davidji99** [comment]
Thanks @ahayworth. Let me get back with more details.

**Comment by robertlaurin** [comment, marked as answer]
Hey @davidji99 are you still stuck on this?

I know we're using the attribute writer successfully to configure custom samplers on various applications in our org.
```rb
OpenTelemetry.tracer_provider.sampler = CustomSampler.new(sampling_probability)
```

**Reply by bala-mane** [reply]
@robertlaurin  
I am getting below error when I use 

`
`<main>': undefined method `sampler=' for #<OpenTelemetry::Internal::ProxyTracerProvider:0x0000000111334e30 @mutex=#<Thread::Mutex:0x0000000111334db8>, @registry={}, @delegate=nil> (NoMethodError)
`

**Reply by jandeschuttere** [reply]
in case someone else runs into this issue, this happens - at least for me - when you call this logic _inside_ the `configure` of Opentelemetry, the sampler should be assigned _after_ the configuration has finished.

**Comment by davidji99** [comment]
Thanks @robertlaurin. Appreciate the sample code.

**Comment by robertlaurin** [comment]
@davidji99 are we ok to close this issue now?

**Comment by arielvalentin** [comment]
@robertlaurin I am going to convert this to a discussion and mark your response as the answer.
